src/tools/schema2graph.ts

```diff
@@ -61,12 +55,15 @@ export class SchemaGraph {
   startNodes: SchemaNode[];
   internalClassIndex = 0;
 
-  constructor(readonly particleSpec: ParticleSpec) {
+  constructor(
+    readonly particleSpec: ParticleSpec,
+    private nameGenerator: (node: SchemaNode, i?: number) => string,
```

Comment:
Nit: `?` means "or undefined". I believe this should actually be `i : number | null`

Reply:
Hmmm, I tried this and it didn't work.

Reply:
Probably because leaving it unassigned at the call site translates to `undefined` (when you don't have a default value). I think `i?` is fine to use, and then in the implementation just use `i == null` - the double-equals form matches against either null or undefined so it's safe.